core/embed/rust/src/ui/model_tt/component/webauthn.rs

```diff
@@ -0,0 +1,235 @@
+    display::{self, Font},
+    geometry::{Insets, Point, Rect},
+    model_tt::{
+        component::{
+            swipe::{Swipe, SwipeDirection},
+            theme, webauthn_icons, ScrollBar,
+        },
+        constant,
+    },
+};
+
+use super::CancelConfirmMsg;
+
+pub enum FidoMsg {
+    Confirmed(u8),
+    Cancelled(u8),
+}
+
+#[derive(Clone, Copy)]
+pub struct FidoPage<T> {
+    app_name: T,
+    account_name: Option<T>,
+}
+
+impl<T> FidoPage<T> {
+    pub fn new(app_name: T, account_name: Option<T>) -> Self {
+        Self {
+            app_name,
+            account_name,
+        }
+    }
+}
+
+pub struct FidoPages<F, T> {
+    /// TODO: find out how to remove this and still allow `T`
+    /// to be passed into the struct.
+    _tmp: T,
+    /// Function/closure that will return appropriate page on demand.
+    get_page: F,
+    /// Number of pages in the flow.
+    page_count: usize,
+}
+
+impl<F, T> FidoPages<F, T>
+where
+    T: AsRef<str> + Copy,
+    F: Fn(usize) -> FidoPage<T>,
+{
+    pub fn new(get_page: F, page_count: usize, tmp: T) -> Self {
+        Self {
+            _tmp: tmp,
+            get_page,
+            page_count,
+        }
+    }
+
+    pub fn get(&self, page_index: usize) -> FidoPage<T> {
+        (self.get_page)(page_index)
+    }
+
+    pub fn count(&self) -> usize {
+        self.page_count
+    }
+}
+
+pub struct FidoPaginatedPage<F, T, U> {
+    page_swipe: Swipe,
+    icon: Child<Image>,
+    pages: FidoPages<F, T>,
+    scrollbar: ScrollBar,
+    content_area: Rect,
+    fade: bool,
+    controls: U,
+}
+
+impl<F, T, U> FidoPaginatedPage<F, T, U>
+where
+    F: Fn(usize) -> FidoPage<T>,
+    T: AsRef<str> + Copy,
+    U: Component<Msg = CancelConfirmMsg>,
+{
+    pub fn new(pages: FidoPages<F, T>, icon_name: Option<T>, controls: U) -> Self {
+        let icon_data = webauthn_icons::get_icon_data(icon_name.as_ref());
+
+        let mut scrollbar = ScrollBar::horizontal();
+        scrollbar.set_count_and_active_page(pages.count(), 0);
+
+        Self {
+            page_swipe: Swipe::horizontal(),
+            icon: Child::new(Image::new(icon_data)),
+            pages,
+            scrollbar,
+            content_area: Rect::zero(),
+            fade: false,
+            controls,
+        }
+    }
+
+    fn on_page_swipe(&mut self, ctx: &mut EventCtx, swipe: SwipeDirection) {
+        // Change the page number.
+        let active_page = match swipe {
+            SwipeDirection::Left if self.scrollbar.has_next_page() => self.active_page() + 1,
+            SwipeDirection::Right if self.scrollbar.has_previous_page() => self.active_page() - 1,
+            _ => self.active_page(), // page did not change
+        };
+        self.scrollbar.go_to(active_page);
+
+        // Disable swipes on the boundaries. Not allowing carousel effect.
+        self.page_swipe.allow_right = self.scrollbar.has_previous_page();
+        self.page_swipe.allow_left = self.scrollbar.has_next_page();
```

Comment:
These doesn't seem to be updated before first swipe, so on the first screen you can probably swipe right even if you shouldn't?

Reply:
Good catch, I am updating that during the object creation - [b557d76](https://github.com/trezor/trezor-firmware/pull/2556/commits/b557d76468ea90567cba981b4288e96502a5b73f)